p4-fusion/commands/change_list.cc

```diff
@@ -108,49 +104,72 @@ void ChangeList::StartDownload(P4API* p4, const BranchSet& branchSet, const int&
 
 	// Flush any remaining files that were smaller in number than the total batch size.
 	// Additionally, signal the batch processing end.
-	Flush(p4, printBatchFiles, printBatchFileData);
+	Flush(p4, printBatchFileData);
 	*downloadJobsCompleted = true;
 	commitCV->notify_all();
 }
 
-void ChangeList::Flush(P4API* p4, std::shared_ptr<std::vector<std::string>> printBatchFiles, std::shared_ptr<std::vector<FileData*>> printBatchFileData)
+void ChangeList::Flush(P4API* p4, const std::shared_ptr<std::vector<FileData*>>& printBatchFileData)
 {
-	// Only perform the batch processing when there are files to process.
-	if (!printBatchFileData->empty())
-	{
-		const PrintResult& printData = p4->PrintFiles(*printBatchFiles);
-		if (printData.HasError())
-		{
-			throw printData.PrintError();
-		}
+	MTR_SCOPE("ChangeList", __func__);
 
-		for (int i = 0; i < printBatchFiles->size(); i++)
-		{
-			printBatchFileData->at(i)->MoveContentsOnceFrom(printData.GetPrintData().at(i).contents);
-		}
+	std::vector<std::string> fileRevisions;
+	for (auto fileData : *printBatchFileData)
+	{
+		fileRevisions.push_back(fileData->GetDepotFile() + "#" + fileData->GetRevision());
+	}
 
-		(*filesDownloaded) += printBatchFiles->size();
+	// Only perform the batch processing when there are files to process.
+	if (fileRevisions.empty())
+	{
+		return;
 	}
 
-	// Ensure the notify_all is called.
-	commitCV->notify_all();
+	long idx = -1;
+	bool flushed = true;
+	PrintResult printResp = p4->PrintFiles(
+	    fileRevisions, [&idx, &printBatchFileData, &flushed]
+	    {
+		    if (idx == -1)
+		    {
+			    idx++;
```

Comment:
It would be great if you had some comments explaining what these callbacks are doing. There are some non-intuitive things (why idx starts at `-1`, etc.)

Reply:
done 👍

Reply:
re: `-1` I believe this can be simplified now (the initial approach was more complex), but that'll require running another test suite for like an hour coz I'm only 98% confident 😀